Node::writeValueArray() compilation broken for std::vector<bool>

###Observed behavior
Node::writeValueArray() compilation broken for std::vector<bool> since c6fe48b. As for `std::vector<double>` everything is working fine thismight be related to std::vector<bool> being a special case. https://en.cppreference.com/w/cpp/container/vector_bool. However, I did not fully understand what is exactly happeing inside the templated code


### Steps to reproduce
use tests\node.cpp file and after
` Node varNode(connection, varId);`
insert 
```
    std::vector<bool> array{true, false};
    varNode.writeValueArray(array);
```
Building results in compilation error:


### Error message
```
C:\source\repos\open62541pp\include\open62541pp\types.hpp(1628): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'opcua::Span<bool>'
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1628): note: 'opcua::Span<bool>::Span(std::initializer_list<bool>) noexcept': cannot convert argument 1 from 'initializer list' to 'std::initializer_list<bool>'
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1628): note: Element '1': no conversion from 'std::vector<bool,std::allocator<bool>>' to 'bool'
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1628): note: while trying to match the argument list '(initializer list)'
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1628): note: the template instantiation context (the oldest one first) is
  C:\source\repos\open62541pp\tests\node.cpp(12): note: see reference to function template instantiation 'void DOCTEST_ANON_TMP_20<Type>(void)' being compiled
          with
          [
              Type=opcua::Server
          ]
  C:\source\repos\open62541pp\tests\node.cpp(78): note: see reference to function template instantiation 'opcua::Node<opcua::Server> &opcua::Node<opcua::Server>::writeValueArray<std::vector<bool,std::allocator<bool>>>(const ArrayLike &)' being compiled
          with
          [
              ArrayLike=std::vector<bool,std::allocator<bool>>
          ]
  C:\source\repos\open62541pp\include\open62541pp/node.hpp(1183): note: see reference to function template instantiation 'opcua::Variant opcua::Variant::fromArray<opcua::VariantPolicy::Reference,std::vector<bool,std::allocator<bool>>&>(ArrayLike)' being compiled
          with
          [
              ArrayLike=std::vector<bool,std::allocator<bool>> &
          ]
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1142): note: see reference to function template instantiation 'opcua::Variant::Variant<std::vector<bool,std::allocator<bool>>&>(opcua::ReferenceTag,T)' being compiled
          with
          [
              T=std::vector<bool,std::allocator<bool>> &
          ]
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1086): note: see reference to function template instantiation 'void opcua::Variant::setValue<std::vector<bool,std::allocator<bool>>&>(T)' being compiled
          with
          [
              T=std::vector<bool,std::allocator<bool>> &
          ]
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1312): note: see reference to function template instantiation 'void opcua::Variant::setValueImpl<std::vector<bool,std::allocator<bool>>&,>(T)' being compiled
          with
          [
              T=std::vector<bool,std::allocator<bool>> &
          ]
C:\source\repos\open62541pp\include\open62541pp\types.hpp(1628): error C2672: 'opcua::Variant::setArray': no matching overloaded function found
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1394): note: could be 'void opcua::Variant::setArray(ArrayLike &&,const UA_DataType &) noexcept'
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1628): note: 'void opcua::Variant::setArray(ArrayLike &&,const UA_DataType &) noexcept': expects 2 arguments - 1 provided
  C:\source\repos\open62541pp\include\open62541pp/types.hpp(1382): note: or       'void opcua::Variant::setArray(ArrayLike &&) noexcept'
  [45/46] Building CXX object tests\CMakeFiles\open62541pp_tests.dir\services_monitoreditem.cpp.obj
```

--- a/include/open62541pp/detail/traits.hpp
+++ b/include/open62541pp/detail/traits.hpp
@@ -1,6 +1,6 @@
 #pragma once
 
-#include <iterator>  // data, size
+#include <iterator>  // data, iterator_traits, size
 #include <type_traits>
 
 namespace opcua::detail {
@@ -31,11 +31,23 @@ struct MemberType<T C::*> {
 template <typename T>
 using MemberTypeT = typename MemberType<T>::type;
 
-template <typename, typename = void>
-struct HasSize : std::false_type {};
+template <typename T>
+using RangeIteratorT = decltype(std::begin(std::declval<T&>()));
 
 template <typename T>
-struct HasSize<T, std::void_t<decltype(std::size(std::declval<T&>()))>> : std::true_type {};
+using RangeValueT = typename std::iterator_traits<RangeIteratorT<T>>::value_type;
+
+template <typename, typename = void>
+struct HasBegin : std::false_type {};
+
+template <typename T>
+struct HasBegin<T, std::void_t<decltype(std::begin(std::declval<T&>()))>> : std::true_type {};
+
+template <typename, typename = void>
+struct HasEnd : std::false_type {};
+
+template <typename T>
+struct HasEnd<T, std::void_t<decltype(std::end(std::declval<T&>()))>> : std::true_type {};
 
 template <typename, typename = void>
 struct HasData : std::false_type {};
@@ -50,32 +62,16 @@ template <typename T>
 struct HasDataPointer<T, std::void_t<decltype(std::data(std::declval<T&>()))>>
     : std::is_pointer<decltype(std::data(std::declval<T&>()))> {};
 
-template <typename T, typename = void>
-struct IsContainer : std::false_type {};
+template <typename, typename = void>
+struct HasSize : std::false_type {};
 
 template <typename T>
-struct IsContainer<
-    T,
-    std::void_t<decltype(std::declval<T>().begin()), decltype(std::declval<T>().end())>>
-    : std::true_type {};
+struct HasSize<T, std::void_t<decltype(std::size(std::declval<T&>()))>> : std::true_type {};
 
 template <typename T>
-inline constexpr bool isContainer = IsContainer<T>::value;
+struct IsRange : std::conjunction<HasBegin<T>, HasEnd<T>> {};
 
 template <typename T>
-struct IsContiguousContainer : std::conjunction<HasSize<T>, HasDataPointer<T>> {};
-
-template <typename T>
-struct IsMutableContainer
-    : std::conjunction<
-          std::is_same<
-              decltype(std::declval<T>().begin()),
-              typename std::remove_reference_t<T>::iterator>,
-          std::is_same<
-              decltype(std::declval<T>().end()),
-              typename std::remove_reference_t<T>::iterator>,
-          std::negation<std::is_same<
-              typename std::remove_reference_t<T>::iterator,
-              typename std::remove_reference_t<T>::const_iterator>>> {};
+struct IsContiguousRange : std::conjunction<IsRange<T>, HasDataPointer<T>, HasSize<T>> {};
 
 }  // namespace opcua::detail

--- a/include/open62541pp/detail/traits.hpp
+++ b/include/open62541pp/detail/traits.hpp
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <iterator>  // data, size
 #include <type_traits>
 
 namespace opcua::detail {
@@ -30,6 +31,25 @@ struct MemberType<T C::*> {
 template <typename T>
 using MemberTypeT = typename MemberType<T>::type;
 
+template <typename, typename = void>
+struct HasSize : std::false_type {};
+
+template <typename T>
+struct HasSize<T, std::void_t<decltype(std::size(std::declval<T&>()))>> : std::true_type {};
+
+template <typename, typename = void>
+struct HasData : std::false_type {};
+
+template <typename T>
+struct HasData<T, std::void_t<decltype(std::data(std::declval<T&>()))>> : std::true_type {};
+
+template <typename, typename = void>
+struct HasDataPointer : std::false_type {};
+
+template <typename T>
+struct HasDataPointer<T, std::void_t<decltype(std::data(std::declval<T&>()))>>
+    : std::is_pointer<decltype(std::data(std::declval<T&>()))> {};
+
 template <typename T, typename = void>
 struct IsContainer : std::false_type {};
 
@@ -42,22 +62,8 @@ struct IsContainer<
 template <typename T>
 inline constexpr bool isContainer = IsContainer<T>::value;
 
-template <typename T, typename = void>
-struct IsContiguousContainer : std::false_type {};
-
 template <typename T>
-struct IsContiguousContainer<
-    T,
-    std::void_t<
-        decltype(std::declval<T>().data()),
-        decltype(std::declval<T>().size()),
-        decltype(std::declval<T>().begin()),
-        decltype(std::declval<T>().end())>> : std::true_type {
-    using value_type = bool;
-    // detect vector<bool>::data() void return type
-    static constexpr bool value = std::is_pointer_v<decltype(std::declval<T>().data())>;
-    using type = std::bool_constant<value>;
-};
+struct IsContiguousContainer : std::conjunction<HasSize<T>, HasDataPointer<T>> {};
 
 template <typename T>
 struct IsMutableContainer

--- a/include/open62541pp/node.hpp
+++ b/include/open62541pp/node.hpp
@@ -1179,7 +1179,8 @@ public:
     /// Write array value to variable node.
     template <typename ArrayLike>
     Node& writeValueArray(const ArrayLike& array) {
-        if constexpr (detail::isRegisteredType<typename ArrayLike::value_type>) {
+        if constexpr (detail::IsContiguousContainer<ArrayLike>::value &&
+                      detail::isRegisteredType<typename ArrayLike::value_type>) {
             // NOLINTNEXTLINE(*-const-cast), variant isn't modified, avoid copy
             writeValue(Variant(const_cast<ArrayLike*>(&array)));
         } else {

--- a/include/open62541pp/node.hpp
+++ b/include/open62541pp/node.hpp
@@ -1179,8 +1179,8 @@ public:
     /// Write array value to variable node.
     template <typename ArrayLike>
     Node& writeValueArray(const ArrayLike& array) {
-        if constexpr (detail::IsContiguousContainer<ArrayLike>::value &&
-                      detail::isRegisteredType<typename ArrayLike::value_type>) {
+        if constexpr (detail::IsContiguousRange<ArrayLike>::value &&
+                      detail::isRegisteredType<detail::RangeValueT<ArrayLike>>) {
             // NOLINTNEXTLINE(*-const-cast), variant isn't modified, avoid copy
             writeValue(Variant(const_cast<ArrayLike*>(&array)));
         } else {

--- a/include/open62541pp/span.hpp
+++ b/include/open62541pp/span.hpp
@@ -11,6 +11,8 @@
 #include <type_traits>
 #include <utility>  // swap
 
+#include "open62541pp/detail/traits.hpp"
+
 namespace opcua {
 
 /**
@@ -27,21 +29,9 @@ namespace opcua {
 template <typename T>
 class Span {
 private:
-    template <typename, typename = void>
-    struct HasSize : std::false_type {};
-
-    template <typename C>
-    struct HasSize<C, std::void_t<decltype(std::size(std::declval<C>()))>> : std::true_type {};
-
-    template <typename, typename = void>
-    struct HasData : std::false_type {};
-
-    template <typename C>
-    struct HasData<C, std::void_t<decltype(std::data(std::declval<C>()))>> : std::true_type {};
-
     template <typename C>
     using EnableIfHasSizeAndData =
-        typename std::enable_if_t<HasSize<C>::value && HasData<C>::value>;
+        typename std::enable_if_t<detail::HasSize<C>::value && detail::HasData<C>::value>;
 
 public:
     // clang-format off

--- a/include/open62541pp/types.hpp
+++ b/include/open62541pp/types.hpp
@@ -1341,7 +1341,7 @@ public:
     template <typename T>
     void assign(const T& value) {
         if constexpr (isArrayType<T>()) {
-            assign(value.begin(), value.end());
+            assign(std::begin(value), std::end(value));
         } else {
             assertIsRegisteredOrConvertible<T>();
             if constexpr (detail::isRegisteredType<T>) {
@@ -1363,7 +1363,7 @@ public:
     template <typename T>
     void assign(const T& value, const UA_DataType& type) {
         if constexpr (isArrayType<T>()) {
-            setArrayCopyImpl(value.begin(), value.end(), type);
+            setArrayCopyImpl(std::begin(value), std::end(value), type);
         } else {
             setScalarCopyImpl(value, type);
         }

--- a/include/open62541pp/types.hpp
+++ b/include/open62541pp/types.hpp
@@ -1306,7 +1306,7 @@ public:
     template <typename T, typename = std::enable_if_t<!std::is_const_v<T>>>
     void assign(T* ptr) noexcept {
         if constexpr (isArrayType<T>()) {
-            using ValueType = typename T::value_type;
+            using ValueType = detail::RangeValueT<T>;
             assertIsRegistered<ValueType>();
             assign(ptr, opcua::getDataType<ValueType>());
         } else {
@@ -1689,7 +1689,7 @@ private:
 
     template <typename T>
     static constexpr bool isArrayType() noexcept {
-        return detail::isContainer<T> && !isScalarType<T>();
+        return detail::IsRange<T>::value && !isScalarType<T>();
     }
 
     template <typename T>

--- a/tests/node.cpp
+++ b/tests/node.cpp
@@ -425,6 +425,12 @@ TEST_CASE_TEMPLATE("Node", T, Server, Client, Async<Client>) {
                 CHECK(varNode.template readValueArray<double>() == array);
             }
         }
+
+        SUBCASE("Array (std::vector<bool>)") {
+            std::vector<bool> array{true, false, true};
+            CHECK_NOTHROW(varNode.writeValueArray(array));
+            CHECK_NOTHROW(varNode.writeValueArray(array.begin(), array.end()));
+        }
     }
 
     SUBCASE("writeDataType/readDataType") {

--- a/tests/traits.cpp
+++ b/tests/traits.cpp
@@ -9,32 +9,54 @@
 
 using namespace opcua;
 
-TEST_CASE_TEMPLATE("IsContiguousContainer true", T, std::vector<int>, std::array<int, 3>, Span<int>, Span<const int>) {
-    CHECK(detail::IsContiguousContainer<T>::value);
-    CHECK(detail::IsContiguousContainer<T&>::value);
-    CHECK(detail::IsContiguousContainer<T&&>::value);
-    CHECK(detail::IsContiguousContainer<const T>::value);
-    CHECK(detail::IsContiguousContainer<const T&>::value);
-    CHECK(detail::IsContiguousContainer<const T&&>::value);
+TEST_CASE("RangeValueT") {
+    CHECK(std::is_same_v<detail::RangeValueT<std::vector<int>>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<std::vector<int>&>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<std::vector<int>&&>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<const std::vector<int>>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<const std::vector<int>&>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<const std::vector<int>&&>, int>);
+
+    CHECK(std::is_same_v<detail::RangeValueT<int[3]>, int>);  // NOLINT(*avoid-c-arrays)
+    CHECK(std::is_same_v<detail::RangeValueT<std::array<int, 3>>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<Span<int>>, int>);
+    CHECK(std::is_same_v<detail::RangeValueT<Span<const int>>, int>);
+
+    CHECK(std::is_same_v<detail::RangeValueT<std::vector<bool>>, bool>);
 }
 
-TEST_CASE_TEMPLATE("IsContiguousContainer false", T, std::list<int>, std::vector<bool>) {
-    CHECK_FALSE(detail::IsContiguousContainer<T>::value);
-    CHECK_FALSE(detail::IsContiguousContainer<T&>::value);
-    CHECK_FALSE(detail::IsContiguousContainer<T&&>::value);
-    CHECK_FALSE(detail::IsContiguousContainer<const T>::value);
-    CHECK_FALSE(detail::IsContiguousContainer<const T&>::value);
-    CHECK_FALSE(detail::IsContiguousContainer<const T&&>::value);
+TEST_CASE_TEMPLATE("IsRange true", T, std::vector<int>, std::array<int, 3>, Span<int>, Span<const int>) {
+    CHECK(detail::IsRange<T>::value);
+    CHECK(detail::IsRange<T&>::value);
+    CHECK(detail::IsRange<T&&>::value);
+    CHECK(detail::IsRange<const T>::value);
+    CHECK(detail::IsRange<const T&>::value);
+    CHECK(detail::IsRange<const T&&>::value);
 }
 
-TEST_CASE_TEMPLATE("IsMutableContainer true", T, std::vector<int>, std::array<int, 3>, Span<int>, std::vector<bool>) {
-    CHECK(detail::IsMutableContainer<T>::value);
-    CHECK(detail::IsMutableContainer<T&>::value);
-    CHECK(detail::IsMutableContainer<T&&>::value);
+TEST_CASE_TEMPLATE("IsRange false", T, bool, int) {
+    CHECK_FALSE(detail::IsRange<T>::value);
+    CHECK_FALSE(detail::IsRange<T&>::value);
+    CHECK_FALSE(detail::IsRange<T&&>::value);
+    CHECK_FALSE(detail::IsRange<const T>::value);
+    CHECK_FALSE(detail::IsRange<const T&>::value);
+    CHECK_FALSE(detail::IsRange<const T&&>::value);
 }
 
-TEST_CASE_TEMPLATE("IsMutableContainer false", T, const std::vector<int>, const std::array<int, 3>, const Span<const int>, const std::vector<bool>) {
-    CHECK_FALSE(detail::IsMutableContainer<T>::value);
-    CHECK_FALSE(detail::IsMutableContainer<T&>::value);
-    CHECK_FALSE(detail::IsMutableContainer<T&&>::value);
+TEST_CASE_TEMPLATE("IsContiguousRange true", T, std::vector<int>, std::array<int, 3>, Span<int>, Span<const int>) {
+    CHECK(detail::IsContiguousRange<T>::value);
+    CHECK(detail::IsContiguousRange<T&>::value);
+    CHECK(detail::IsContiguousRange<T&&>::value);
+    CHECK(detail::IsContiguousRange<const T>::value);
+    CHECK(detail::IsContiguousRange<const T&>::value);
+    CHECK(detail::IsContiguousRange<const T&&>::value);
+}
+
+TEST_CASE_TEMPLATE("IsContiguousRange false", T, std::list<int>, std::vector<bool>) {
+    CHECK_FALSE(detail::IsContiguousRange<T>::value);
+    CHECK_FALSE(detail::IsContiguousRange<T&>::value);
+    CHECK_FALSE(detail::IsContiguousRange<T&&>::value);
+    CHECK_FALSE(detail::IsContiguousRange<const T>::value);
+    CHECK_FALSE(detail::IsContiguousRange<const T&>::value);
+    CHECK_FALSE(detail::IsContiguousRange<const T&&>::value);
 }

